vault/namespaces_openbao.go

```diff
@@ -0,0 +1,475 @@
+			return handleError(err)
+		}
+
+		return nil, nil
+	}
+}
+
+// handleNamespacesPatch handles the "/sys/namespace/<path>" endpoints to update a namespace
+func (b *SystemBackend) handleNamespacesPatch() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		path := data.Get("path").(string)
+		imetadata, ok := data.GetOk("custom_metadata")
+		var metadata map[string]string
+		if ok {
+			metadata = imetadata.(map[string]string)
+		}
+
+		// Update the namespace
+		if err := b.Core.namespaceStore.PatchNamespace(ctx, path, metadata); err != nil {
+			return handleError(err)
+		}
+
+		return nil, nil
+	}
+}
+
+func (b *SystemBackend) handleNamespacesDelete() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		path := data.Get("path").(string)
+
+		if err := b.Core.namespaceStore.DeleteNamespace(ctx, path); err != nil {
+			return handleError(err)
+		}
+		return nil, nil
+	}
+}
+
+// NamespaceStore is used to provide durable storage of namespace
+type NamespaceStore struct {
+	core    *Core
+	aclView BarrierView
+
+	// This is used to ensure that writes to the store (acl) or to the egp
+	// path tree don't happen concurrently. We are okay reading stale data so
+	// long as there aren't concurrent writes.
+	modifyLock *sync.RWMutex
+
+	// logger is the server logger copied over from core
+	logger hclog.Logger
+}
+
+// NamespaceEntry is used to store a namespace by name
+type NamespaceEntry struct {
+	namespace.Namespace
+}
+
+// NewNamespaceStore creates a new NamespaceStore that is backed
+// using a given view. It used used to durable store and manage named namespace.
+func NewNamespaceStore(ctx context.Context, core *Core, baseView BarrierView, system logical.SystemView, logger hclog.Logger) (*NamespaceStore, error) {
+	ps := &NamespaceStore{
+		aclView:    baseView.SubView(namespaceSubPath),
+		modifyLock: new(sync.RWMutex),
+		logger:     logger,
+		core:       core,
+	}
+
+	// initialize the namespace store to have the root namespace
+	err := ps.SetNamespace(ctx, namespace.RootNamespaceID, map[string]string{}, true)
+	return ps, err
+}
+
+// setupNamespaceStore is used to initialize the namespace store
+// when the vault is being unsealed.
+func (c *Core) setupNamespaceStore(ctx context.Context) error {
+	// Create the Namespace store
+	var err error
+	sysView := &dynamicSystemView{core: c}
+	psLogger := c.baseLogger.Named("namespace")
+	c.AddLogger(psLogger)
+	c.namespaceStore, err = NewNamespaceStore(ctx, c, c.systemBarrierView, sysView, psLogger)
+
+	return err
+}
+
+// teardownNamespaceStore is used to reverse setupNamespaceStore
+// when the vault is being sealed.
+func (c *Core) teardownNamespaceStore() error {
+	c.namespaceStore = nil
+	return nil
+}
+
+func (ps *NamespaceStore) invalidate(ctx context.Context, path string) error {
```

Comment:
```suggestion
func (ns *NamespaceStore) invalidate(ctx context.Context, path string) error {
```

How about `ns` rather than `ps` for NamespaceStore?